---


---

<pre><code>library(GenomicRanges)
library(BiocParallel)
library(ggplot2)
#加载数据
load("chromosome_lengths.RData")
load("utr_3_prime_regions_p.RData")
load("utr_3_prime_regions_m.RData")

#封装函数
run_utr_enrichment &lt;- function(
  motif_name,
  n_simulations,
  sequences_per_simulation,
  sequence_length,
  chromosome_lengths,
  utr_3_prime_regions_p,
  utr_3_prime_regions_m,
  x_observed,
  y_observed,
  n_cores = 4
) {
  #计算有效长度
  utr_3_prime_regions_p$effective_end &lt;- utr_3_prime_regions_p$end - sequence_length + 1
  utr_3_prime_regions_m$effective_end &lt;- utr_3_prime_regions_m$end - sequence_length + 1
  
  bad_p &lt;- which(utr_3_prime_regions_p$effective_end &lt; utr_3_prime_regions_p$start)
  bad_m &lt;- which(utr_3_prime_regions_m$effective_end &lt; utr_3_prime_regions_m$start)
  cat("bad_p:", length(bad_p), "\n")
  cat("bad_m:", length(bad_m), "\n")  
  
  utr_3_prime_regions_p &lt;- utr_3_prime_regions_p[utr_3_prime_regions_p$effective_end &gt;= utr_3_prime_regions_p$start, ]
  utr_3_prime_regions_m &lt;- utr_3_prime_regions_m[utr_3_prime_regions_m$effective_end &gt;= utr_3_prime_regions_m$start, ] 

  utr_p_gr &lt;- GRanges(
    seqnames = utr_3_prime_regions_p$chromosome,
    ranges = IRanges(start = utr_3_prime_regions_p$start,
                     end = utr_3_prime_regions_p$effective_end),
    strand = utr_3_prime_regions_p$strand
  )
  utr_m_gr &lt;- GRanges(
    seqnames = utr_3_prime_regions_m$chromosome,
    ranges = IRanges(start = utr_3_prime_regions_m$start,
                     end = utr_3_prime_regions_m$effective_end),
    strand = utr_3_prime_regions_m$strand
  )
  
  chromosome_data &lt;- data.frame(
    chromosome = names(chromosome_lengths),
    length = chromosome_lengths
  )
 
  cat("Starting simulations for motif:", motif_name, "...\n")
  cat("&gt;&gt; Simulating", n_simulations, "times with", sequences_per_simulation, "sequences per run\n")
  sim_results &lt;- bplapply(1:n_simulations, FUN = function(i) {
    if (i %% 1000 == 0) cat("  Simulation", i, "of", n_simulations, "\n")
    random_chromosomes &lt;- sample(chromosome_data$chromosome, sequences_per_simulation, replace = TRUE)
    random_strand &lt;- sample(c("+", "-"), sequences_per_simulation, replace = TRUE)
    random_positions &lt;- sapply(random_chromosomes, function(chr) {
      sample(1:(chromosome_lengths[chr] - sequence_length + 1), 1)
    })
    
    random_gr &lt;- GRanges(
      seqnames = random_chromosomes,
      ranges = IRanges(start = random_positions,
                       end = random_positions + sequence_length - 1),
      strand = random_strand
    )
    
    sense_hits &lt;- countOverlaps(random_gr, utr_p_gr, type = "within")
    antisense_hits &lt;- countOverlaps(random_gr, utr_m_gr, type = "within")
    
    list(sense = sum(sense_hits &gt; 0), antisense = sum(antisense_hits &gt; 0))
  }, BPPARAM = MulticoreParam(workers = n_cores))
  
  utr_sense_count &lt;- sapply(sim_results, `[[`, "sense")
  utr_antisense_count &lt;- sapply(sim_results, `[[`, "antisense")
  
  save(utr_sense_count, file = paste0("utr_sense_count_", motif_name, ".RData"))
  save(utr_antisense_count, file = paste0("utr_antisense_count_", motif_name, ".RData"))
  

  plot_density_with_arrow &lt;- function(sim_data, observed_value, title, color = "blue") {
  df &lt;- data.frame(counts = sim_data)
  
  p &lt;- ggplot(df, aes(x = counts)) +
    geom_density(fill = color, alpha = 0.4, color = color, size = 1.2) +
    
    annotate("segment",
             x = observed_value, xend = observed_value,
             y = 0, yend = 0.02, 
             arrow = arrow(length = unit(0.25, "cm")), 
             color = "red", size = 1.2) +
    
    annotate("text",
             x = observed_value,
             y = 0.025,
             label = paste0("observed_value: ", observed_value),
             color = "red",
             angle = 90,
             vjust = -0.5,
             size = 3.5) +
    
    labs(title = title,
         x = "Occurrences in 3'UTR sense",
         y = "Probability") +
    
    theme_minimal()
  
  return(p)
}

  p_sense &lt;- plot_density_with_arrow(utr_sense_count, x_observed, motif_name)
  p_antisense &lt;- plot_density_with_arrow(utr_antisense_count, y_observed, motif_name)
  ggsave(filename = paste0("utr_sense_density_", motif_name, ".png"), plot = p_sense)
  ggsave(filename = paste0("utr_antisense_density_", motif_name, ".png"), plot = p_antisense)
  
  return(list(sense = utr_sense_count, antisense = utr_antisense_count, p_sense = p_sense, p_antisense = p_antisense))
}

motif_info_df &lt;- data.frame(
  motif_name = c("ATTTTGCTT", "ATTATGCTT","ATATTGCTT","AATTTGCTT"),  
  sequences_per_simulation = c(93113, 38387, 45061, 66106),
  x_observed = c(358, 140, 142, 243),
  y_observed = c(354, 133, 146, 221),
  stringsAsFactors = FALSE
)



results_list &lt;- list()
for (i in seq_len(nrow(motif_info_df))) {
  motif &lt;- motif_info_df$motif_name[i]
  seq_num &lt;- motif_info_df$sequences_per_simulation[i]
  x_obs &lt;- motif_info_df$x_observed[i]
  y_obs &lt;- motif_info_df$y_observed[i]
  
  cat("Running motif:", motif, "\n")
  start_i &lt;- Sys.time()
  res &lt;- run_utr_enrichment(
    motif_name = motif,
    n_simulations = 10000,
    sequences_per_simulation = seq_num,
    sequence_length = 9,
    chromosome_lengths = chromosome_lengths,
    utr_3_prime_regions_p = utr_3_prime_regions_p,
    utr_3_prime_regions_m = utr_3_prime_regions_m,
    x_observed = x_obs,
    y_observed = y_obs,
    n_cores = 4
  )
  end_i &lt;- Sys.time()
  cat("==&gt; Finished motif:", motif, "in", round(difftime(end_i, start_i, units = "mins"), 2), "minutes\n")
  results_list[[motif]] &lt;- res
}

save(results_list, file = "all_motif_enrichment_results.RData")
</code></pre>

