---


---

<h1 id="sra-toolkit（2.11.1）"><span class="prefix"></span><span class="content">SRA-toolkit（2.11.1）</span><span class="suffix"></span></h1>
<pre><code>tar zxvf /disk1/shares/sratoolkit.2.11.1-buntu64.tar.gz -C ~/Biosofts 
#我是在老师的服务器下进行本次实验，所以直接沿用之前上课下载好的软件 
#该软件以预编译安装方式安装
cd Biosofts/sratoolkit.2.11.1-buntu64/bin/ 
./prefetch -h 
echo 'export PATH=~/Biosofts/sratoolkit.2.11.1-buntu64/bin:$PATH' &gt;&gt; ~/.bashrc
source ~/.bashrc 
cd ~ prefetch -h
</code></pre>
<h1 id="prefetch批量下载sra数据并转化为fastq.gz的脚本"><span class="prefix"></span><span class="content">prefetch批量下载sra数据并转化为fastq.gz的脚本</span><span class="suffix"></span></h1>
<pre><code>##nano prefetch.sh
#!/bin/bash
for i in 7 8
do
prefetch SRR1845592${i}
done
for i in 1 2
do
prefetch SRR1845593${i}
done
for j in SRR*/*sra
do
fasterq-dump --split-files $j
done
##ctrl+x 保存文件并退出
##chmod +x prefetch.sh
##./prefetch.sh
</code></pre>
<h1 id="fastqc-（0.11.7）"><span class="prefix"></span><span class="content">Fastqc （0.11.7）</span><span class="suffix"></span></h1>
<pre><code>#先下载java环境
cd ~/Biosofts
wget http://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.7.zip
mkdir ~/Biosofts/fastqc
unzip /disk1/shares/fastqc_v0.11.7.zip -d ~/Biosofts/
chmod +x ~/Biosofts/FastQC/fastqc
~/Biosofts/FastQC/fastqc -h
#配置环境变量
echo 'export PATH=~/Biosofts/FastQC:$PATH'&gt;&gt;~/.bashrc
source ~/.bashrc
fastqc -h
#在python2环境下用conda下载multiqc
conda create --name python2 python=2.7 -c https://mirrors.ustc.edu.cn/anaconda/cloud/bioconda/ -y
conda activate python2
#选择镜像下载
conda install multiqc -c https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/bioconda
</code></pre>
<h1 id="cutadapt-（1.18）"><span class="prefix"></span><span class="content">cutadapt （1.18）</span><span class="suffix"></span></h1>
<pre><code>#需先创建一个python3虚拟环境，并进入环境
conda create -n python3 python=3.7
conda activate python3
conda install -c bioconda cutadapt
</code></pre>
<h1 id="trim-galore（0.45）"><span class="prefix"></span><span class="content">Trim Galore（0.45）</span><span class="suffix"></span></h1>
<pre><code>#需要先下载fastqc和cutadapt
curl -fsSL https://github.com/FelixKrueger/TrimGalore/archive/0.6.6.tar.gz -o trim_galore.tar.gz
tar zxvf trim_galore.tar.gz -C ~/Biosofts
cd Biosofts/TrimGalore-0.4.5
./trim_galore
echo 'export PATH=~/Biosofts/TrimGalore-0.4.5:$PATH' &gt;&gt; ~/.bashrc
source ~/.bashrc 
cd ~ 
</code></pre>
<h1 id="bowtie2"><span class="prefix"></span><span class="content">Bowtie2</span><span class="suffix"></span></h1>
<pre><code>conda install -y bowtie2
</code></pre>
<h1 id="samtools"><span class="prefix"></span><span class="content">samtools</span><span class="suffix"></span></h1>
<pre><code>conda insall samtools
</code></pre>
<h1 id="featurecounts"><span class="prefix"></span><span class="content">FeatureCounts</span><span class="suffix"></span></h1>
<pre><code>wget https://jaist.dl.sourceforge.net/project/subread/subread-1.6.0/subread-1.6.0-Linux-x86_64.tar.gz
tar -zxvf subread-1.6.0-Linux-x86_64.tar.gz -C ~/Biosofts
cd Biosofts/subread-2.0.2-Linux-x86_64/bin
./featureCounts
echo 'export PATH=~/Biosofts/subread-2.0.2-Linux-x86_64/bin:$PATH' &gt;&gt; ~/.bashrc
source ~/.bashrc
cd ~
</code></pre>
<h1 id="质控"><span class="prefix"></span><span class="content">质控</span><span class="suffix"></span></h1>
<pre><code>fastqc *.fastq.gz -o ./
multiqc .
</code></pre>
<h1 id="修剪"><span class="prefix"></span><span class="content">修剪</span><span class="suffix"></span></h1>
<pre><code>##nano trimgalore.sh
#!/bin/bash
COUNTER=0
#为文件创建列表
for f in *.fastq.gz ; do
size=${#f}
just_name=${f:0:$size - 11} #只保留run的名字
array_names[$COUNTER]=$just_name #把文件名加入数组
let COUNTER=COUNTER+1
done
#删除重复文件名并排序
sorted_unique_ids=($(echo "${array_names[@]}" | tr ' ' '\n' | sort -u | tr '\n' ' '))
#指定文件储存地方
R1="_1.fastq.gz"
R2="_2.fastq.gz"
for i in "${sorted_unique_ids[@]}" ; do
pair1="${i}${R1}"
pair2="${i}${R2}"
trim_galore -q 14 --paired $pair1 $pair2 --stringency 3 --length 25 --phred33 --gzip -o trim_galore_files/
echo "${i}: Has been completed"
done
##ctrl+x 保存文件并退出
##chmod +x trimgalore.sh
##./trimgalore.sh
</code></pre>
<h1 id="比对"><span class="prefix"></span><span class="content">比对</span><span class="suffix"></span></h1>
<pre><code>#自建立参考基因组索引
#下载参考基因组*.fna.gz
gunzip GCF_000195955.2_ASM19595v2_genomic.fna.gz
cat *.fna &gt; mm10.fa #修改名字，使其简洁
bowtie2-build mm10.fa mm10
##nano Bowtie2.sh 
#!/bin/bash 
RAW_FILES=~/mywork/trim_galore_files/ #输入文件的位置
ALIGNED_FILES=~/mywork/bowtie2_aligned/ #输出文件的位置
R1="_1_val_1.fq.gz"
R2="_2_val_2.fq.gz"
run_name=SRR1845592
for f in 7 8
do
pair1="${RAW_FILES}${run_name}${f}${R1}" #样本文件双端之一
pair2="${RAW_FILES}${run_name}${f}${R2}" #样本文件双端之一
filename="${ALIGNED_FILES}${run_name}${f}.sam"
filename_bam="${ALIGNED_FILES}${run_name}${f}.bam"
filename_sorted_bam="${ALIGNED_FILES}${run_name}${f}.sort.bam"
bowtie2 -p 8 -x /disk1/202031107010076/mywork/trim_galore_files/mm10 --phred33 -N 1 --very-sensitive-local -1 $pair1 -2 $pair2 -S $filename #比对
samtools view -@ 28 -bS -o $filename_bam $filename #sam文件转为bam
samtools sort $filename_bam &gt; $filename_sorted_bam #排序
samtools index $filename_sorted_bam #为sort.bam文件建立索引
echo "${run_name}${f}: Has been completed"
done
##ctrl+x 保存文件并退出
##chmod +x Bowtie2.sh
##./Bowtie2.sh
</code></pre>
<h1 id="计数"><span class="prefix"></span><span class="content">计数</span><span class="suffix"></span></h1>
<pre><code>featureCounts -T 10 -a GCF_000195955.2_ASM19595v2_genomic.gtf
-o ~/mywork/cound.txt
-p -M -t gene -g gene_id -s 2 
~/mywork/bowtie2_aligned/*.sort.bam
</code></pre>
<h1 id="deseq差异分析"><span class="prefix"></span><span class="content">DESeq差异分析</span><span class="suffix"></span></h1>
<pre><code>if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("DESeq2")
library(DESeq2) #下载并载入DESeq包
setwd("D:/") #指定文件读取位置
#读入counts文件
data &lt;- read.table("counts.txt", header=TRUE, skip=1, row.names=1) 
  #说明第一行包含了变量的逻辑型变量，跳过一行，行名是第一列的名称
colnames(data) &lt;- gsub(".sort.bam", "", colnames(data), fixed=TRUE)
colnames(data) &lt;- gsub("X.disk1.202031107010076.mywork.bowtie2_aligned.", "", colnames(data), fixed=TRUE)
  #将样本名替换为SRR*
countdata &lt;- data[ , 6:ncol(data)] #删减2到5列
countdata &lt;- countdata[rowMeans(countdata)&gt;1,] #去除表达量过低的基因
#提前构建信息
condition &lt;- factor(c(rep("hypoxic",2),rep("exponential",2))) 
  #构建缺氧和对数生长两种因子
coldata &lt;- data.frame(row.names=colnames(countdata), condition)
  #构建以样本名为行名进行因子分类的列表
#开始进行差异表达分析
dds &lt;- DESeqDataSetFromMatrix(countData = countData, colData = colData, design = ~ condition)
#标准化
dds1 &lt;- DESeq(dds, fitType = 'mean', minReplicatesForReplace = 7, parallel = FALSE) 
res &lt;- results(dds1) #将结果用result()函数来获取
summary(res) #查看结果
res1 &lt;- data.frame(res, stringsAsFactors = FALSE, check.names = FALSE) #res格式转化为表格形式
res1 &lt;- res1[order(res1$pvalue, res1$log2FoldChange, decreasing = c(FALSE, TRUE)), ]
  #依次按照pvalue值log2FoldChange值进行排序
#获取padj小于0.05，表达倍数取以2为对数后大于1或者小于-1的差异表达基因。
res1_up&lt;- res1[which(res1$log2FoldChange &gt;= 1 &amp; res1$padj &lt; 0.05),] # 表达量显著上升的基因
res1_down&lt;- res1[which(res1$log2FoldChange &lt;= -1 &amp; res1$padj &lt; 0.05),] # 表达量显著下降的基因
res1_total &lt;- rbind(res1_up,res1_down)
#可视化
library(ggplot2)
#PCA 主成分分析图
dds_rlog &lt;- rlog(dds, blind=FALSE)
plotPCA(dds_rlog, intgroup="condition", ntop=100) +
  theme_bw() + # 修改主体
  geom_point(size=3) + # 增加点大小
  scale_y_continuous(limits=c(-5, 5)) +
  ggtitle(label="Principal Component Analysis (PCA)",
          subtitle="Top 100 most variable genes")
#热图
install.packages("pheatmap")
install.packages("RColorBrewer")
library(pheatmap)
library(RColorBrewer)
dds_rlog2 &lt;- rlog(dds, blind=FALSE)
df &lt;- countdata[intersect(rownames(countdata),rownames(res1_total)),] #在原表达矩阵中找到差异表达基因
mat &lt;- assay(dds_rlog2[row.names(df)])[1:40, ] #获得前40差异基因
  #构建列注释信息
annotation_col &lt;- data.frame(
  condition=factor(colData(dds_rlog2)$condition),
  row.names=rownames(colData(dds_rlog)))
  #修改颜色
ann_colors &lt;- list(
  Group=c(LoGlu="lightblue", HiGlu="darkorange"),
  Replicate=c(Rep1="darkred", Rep2="forestgreen"))
  #画图
pheatmap(mat=mat, 
         color = colorRampPalette(c("navy", "white", "firebrick3"))(50), 
         scale="row", 
         annotation_col=annotation_col,
         annotation_colors=ann_colors,
         fontsize=6.5, 
         cellwidth=55, 
         show_colnames=T,
         clustering_method = "complete",
         clustering_distance_rows = "euclidean",
         treeheight_row = 35, treeheight_col = 35,border_color = "black")
#火山图
install.packages("dplyr")
library(dplyr)
vol_data &lt;- data.frame(gene=row.names(res), pval=-log10(res$padj), lfc=res$log2FoldChange)
  #去除空值
vol_data &lt;- na.omit(vol_data)
  #设定上调与下调颜色
vol_data &lt;- mutate(vol_data, color=case_when(
  vol_data$lfc &gt; 0 &amp; vol_data$pval &gt; 1.3 ~ "Increased",
  vol_data$lfc &lt; 0 &amp; vol_data$pval &gt; 1.3 ~ "Decreased",
  vol_data$pval &lt; 1.3 ~ "nonsignificant"))
vol &lt;- ggplot(vol_data, aes(x=lfc, y=pval, color=color))
  #改图例
vol + ggtitle(label="Volcano Plot", subtitle="Colored by fold-change direction") +
  geom_point(size=2.5, alpha=0.8, na.rm=T) +
  scale_color_manual(name="Directionality",
                     values=c(Increased="#008B00", Decreased="#CD4F39",
                              nonsignificant="darkgray")) +
  theme_bw(base_size=14) +
  theme(legend.position="right") +
  xlab(expression(log[2]("hypoxic" / "exponential"))) +
  ylab(expression(-log[10]("adjusted p-value"))) +
  geom_hline(yintercept=1.3, colour="darkgrey") +
  scale_y_continuous(trans="log1p")
#MA图
plotMA(res, ylim=c(-5, 5))
</code></pre>

