# 统计主要转录组和非转录组的碱基数
1.  **提取**：从 GTF 取出感兴趣的转录本坐标
2.  **转格式**：变成 BED
3.  **去重合并**：避免重复计算   
4.  **统计** transcript 碱基数
5.  **生成基因组区间**    
6.  **相减** 得到非 transcript 区域  
7.  **统计**非 transcript 碱基数
8.  **统计**5'UTR 长度
9. **统计**CDS 长度
10.   **统计** 3'UTR 长度

这样 transcript + non-transcript = 基因组总长度（只要 BED 区间完全覆盖全基因组且不重叠）。
```
#!/bin/bash
GTF="protein_coding_mane_transcript.gtf"
FIVEGTF="5utr_mane.gtf"
THREEGTF="3utr_mane.gtf"
CGTF="cds_mane.gtf"
EXONGTF="mane_exons.gtf"
FASTA="Homo_sapiens.GRCh38.dna.primary_assembly.fa"

cut -f1,4,5 $GTF | awk 'BEGIN{OFS="\t"} {print $1, $2-1, $3}' > mane_pc_transcript.bed
bedtools sort -i mane_pc_transcript.bed > mane_pc_transcript_merged.bed
awk '{sum += $3 - $2} END {print "Transcript region length:", sum}' mane_pc_transcript_merged.bed

samtools faidx $FASTA
cut -f1,2 ${FASTA}.fai | awk 'BEGIN{OFS="\t"} {print $1, 0, $2}' > genome.bed
bedtools subtract -a genome.bed -b mane_pc_transcript_merged.bed > non_transcript.bed
awk '{sum += $3 - $2} END {print "Non-transcript region length:", sum}' non_transcript.bed

cut -f1,4,5 $FIVEGTF | awk 'BEGIN{OFS="\t"} {print $1, $2-1, $3}' | bedtools sort -i - | awk '{sum += $3 - $2} END {print "5 prime UTR length:", sum}'

cut -f1,4,5 $CGTF | awk 'BEGIN{OFS="\t"} {print $1, $2-1, $3}' | bedtools sort -i - | awk '{sum += $3 - $2} END {print "CDS length:", sum}'

cut -f1,4,5 $THREEGTF | awk 'BEGIN{OFS="\t"} {print $1, $2-1, $3}' | bedtools sort -i - | awk '{sum += $3 - $2} END {print "3 prime UTR length:", sum}'

cut -f1,4,5 $EXONGTF | awk 'BEGIN{OFS="\t"} {print $1, $2-1, $3}' | bedtools sort -i - | awk '{sum += $3 - $2} END {print "Exon length:", sum}'

cut -f1,4,5 $FIVEGTF | awk 'BEGIN{OFS="\t"} {print $1, $2-1, $3}' | bedtools sort -i -  > 5utr_merged.bed
cut -f1,4,5 $CGTF | awk 'BEGIN{OFS="\t"} {print $1, $2-1, $3}' | bedtools sort -i - > cds_merged.bed
cut -f1,4,5 $THREEGTF | awk 'BEGIN{OFS="\t"} {print $1, $2-1, $3}' | bedtools sort -i - > 3utr_merged.bed

cat 5utr_merged.bed cds_merged.bed 3utr_merged.bed | bedtools sort -i - | awk '{sum += $3 - $2} END {print "CDS+UTR length:", sum}'

bedtools intersect -a 5utr_merged.bed -b cds_merged.bed | awk '{sum += $3 - $2} END {print "over of 5UTR and CDS:", sum}'

bedtools intersect -a cds_merged.bed -b 3utr_merged.bed | awk '{sum += $3 - $2} END {print "over of CDS and 3UTR:", sum}'
```
### 结果
Transcript region length: 1134954587
Non-transcript region length: 1964796131
5 prime UTR length: 3798636
CDS length: 33373991
3 prime UTR length: 33367231
Exon length: 70467196

Transcript region length: 1134954587
Non-transcript region length: 1964796131
5 prime UTR length: 3798636
CDS length: 33373991
3 prime UTR length: 33367231
Exon length: 70467196
CDS+UTR length: 70410805

Transcript region length: 1134954587
Non-transcript region length: 1964796131
5 prime UTR length: 3798636
CDS length: 33373991
3 prime UTR length: 33367231
Exon length: 70467196
CDS+UTR length: 70410805
over of 5UTR and CDS: 10765
over of CDS and 3UTR: 109181

<!--stackedit_data:
eyJoaXN0b3J5IjpbLTE1Njg2OTYzNTUsNzI2ODEzMjQ2LDk5Nz
M1OTQ4NCwtMzk4NjQ5MzE2LC0xMDM2MDk5NjM3LC0xMDk3NzE0
NDI4LDE3NzM2MjM1MzgsLTQ1NDYzNSwxNTgyNjM5OTEwLDEwNj
AxMDgzMTUsMTg5OTkxNzU0MywxNTQ3MjA4Mjk1LDEzODU4NjMx
NTUsMTgxMzg5ODc4NV19
-->