<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>（二）基因表达定量</title>
  <link rel="stylesheet" href="https://stackedit.cn/style.css" />
</head>

<body class="stackedit">
  <div class="stackedit__html"><h1 id="1.下载数据---------2.基因表达定量"><span class="prefix"></span><span class="content">1.下载数据---------2.基因表达定量</span><span class="suffix"></span></h1>
<h2 id="fastq文件处理：cellranger"><span class="prefix"></span><span class="content">fastq文件处理：cellranger</span><span class="suffix"></span></h2>
<h2 id="安装"><span class="prefix"></span><span class="content">安装</span><span class="suffix"></span></h2>
<p><a href="https://www.10xgenomics.com/support/software/cell-ranger/downloads">Download Cell Ranger - Official 10x Genomics Support</a><br>
<img src="https://raw.githubusercontent.com/ZYyli/bioinfosoft_pictures/master/imgs/2024-03-12/XAfV8AXbIBQzeil0.jpeg" alt="输入图片说明"><br>
下载完后查看md5sum值，看下载是否完整。<br>
<code>#解压 tar -zxvf cellranger-7.2.0.tar.gz</code><br>
<code>#修改环境变量 echo 'export PATH=/cellranger-7.2.0/:$PATH' &gt;&gt; ~/.bashrc</code><br>
<code>#更新配置文件 source ~/.bashrc cellranger count --help</code></p>
<h2 id="使用"><span class="prefix"></span><span class="content">使用</span><span class="suffix"></span></h2>
<h3 id="建立参考基因组经过试验，后决定还是使用官方给定参考基因版本"><span class="prefix"></span><span class="content">建立参考基因组(经过试验，后决定还是使用官方给定参考基因版本)</span><span class="suffix"></span></h3>
<p>10x官方有人和小鼠的参考基因组，但我这里采用最新人类参考基因组<br>
<img src="https://raw.githubusercontent.com/ZYyli/bioinfosoft_pictures/master/imgs/2024-03-12/iV1bwhDfgivcnMFL.png" alt="输入图片说明"><br>
下载参考基因组(fasta)和注释基因(GTF)<br>
<code>wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/GCF_000001405.40_GRCh38.p14_genomic.fna.gz wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/GCF_000001405.40_GRCh38.p14_genomic.gtf.gz</code></p>
<h4 id="cellranger-mkgtf-对gtf文件进行过滤"><span class="prefix"></span><span class="content">Cellranger mkgtf 对GTF文件进行过滤</span><span class="suffix"></span></h4>
<p>10x单细胞使用的polydT进行RNA逆转录，只能测到带有polyA尾的RNA序列，所以我们需要从GTF文件中过滤掉non-polyA的基因。Cellranger的mkgtf命令可以对GTF文件进行过滤，–attribute=gene_biotype:protein_coding<br>
<code>cellranger mkgtf GCF_000001405.40_GRCh38.p14_genomic.gtf GCF_000001405.40_GRCh38.p14_genomic.filtered.gtf --attribute=gene_biotype:protein_coding --attribute=gene_biotype:lincRNA --attribute=gene_biotype:antisense --attribute=gene_biotype:IG_LV_gene --attribute=gene_biotype:IG_V_gene --attribute=gene_biotype:IG_V_pseudogene --attribute=gene_biotype:IG_D_gene --attribute=gene_biotype:IG_J_gene --attribute=gene_biotype:IG_J_pseudogene --attribute=gene_biotype:IG_C_gene --attribute=gene_biotype:IG_C_pseudogene --attribute=gene_biotype:TR_V_gene --attribute=gene_biotype:TR_V_pseudogene --attribute=gene_biotype:TR_D_gene --attribute=gene_biotype:TR_J_gene --attribute=gene_biotype:TR_J_pseudogene --attribute=gene_biotype:TR_C_gene</code><br>
<code>#这样得到的Homo_sapiens.GRCh38.ensembl.filtered.gtf结果中就不包含gene_biotype:pseudogene这部分</code></p>
<h4 id="cellranger-mkref命令构建"><span class="prefix"></span><span class="content">cellranger mkref命令构建</span><span class="suffix"></span></h4>
<p>–genome：生成索引的目录<br>
–fasta：基因组序列<br>
–genes：基因注释文件（gtf格式）<br>
<code>cellranger mkref --genome=hg --nthreads=12 --fasta=GCF_000001405.40_GRCh38.p14_genomic.fna --genes=GCF_000001405.40_GRCh38.p14_genomic.gtf</code><br>
10X官方参考基因组<br>
<a href="https://www.10xgenomics.com/support/software/space-ranger/downloads#reference-downloads">Download Space Ranger - Official 10x Genomics Support</a></p>
<h2 id="cellranger-count定量"><span class="prefix"></span><span class="content">cellranger count定量</span><span class="suffix"></span></h2>
<h4 id="fastq文件改名"><span class="prefix"></span><span class="content">fastq文件改名</span><span class="suffix"></span></h4>
<p><code>cat SRR.txt | while read i ;do (mv ${i}_1*.gz ${i}_S1_L001_R1_001.fastq.gz;mv ${i}_2*.gz ${i}_S1_L001_R2_001.fastq.gz);done</code></p>
<h4 id="count函数解释"><span class="prefix"></span><span class="content">count函数解释</span><span class="suffix"></span></h4>
<p><code>sample='85 86 87 88 89 90' for i in $sample do cellranger count --id=SRR101343$i_test_out --transcriptome=refdata-gex-GRCh38-2020-A --fastqs=fq --sample=SRR101343$i --nosecondary done</code><br>
<code>#id指定输出文件存放目录名 #transcriptome指定与CellRanger兼容的参考基因组 #fastqs指定mkfastq或者自定义的测序文件 #sample要和fastq文件的前缀中的sample保持一致，作为软件识别的标志 #expect-cells指定复现的细胞数量，这个要和实验设计结合起来 #nosecondary 只获得表达矩阵，不进行后续的降维、聚类和可视化分析(反正后续要走Seurat，为了节省计算资源，建议加上)</code></p>
<h3 id="结果解读"><span class="prefix"></span><span class="content">结果解读</span><span class="suffix"></span></h3>
</div>
</body>

</html>
